### PR TITLE
Replaces dibs with Bay's ticket system.

### PR DESCRIPTION
**Pull request by mustafakalash**
image
You guys want this?

- A user ahelping creates a ticket. Any further ahelps while that ticket is open will go to that ticket (and either adminhelp or pm an admin assigned to the ticket, if one has taken it). 
- A user can close their own ticket up until an admin takes it.
- Admins can take tickets either manually or just by replying to the PM.
- An admin taking a ticket notifies other admins.
- If another admin attempts to take or respond to an assigned ticket, they receive a notification asking if they would like to join the ticket or cancel (any number of admins can join a ticket).
- When an admin finishes with a ticket, they must close it -- both for logistics and closure for the user.
- If there is an open ticket assigned to an active admin, round end is automatically delayed (unassigned tickets or tickets assigned to disconnected or afk admins do not delay). Round end will automatically continue once all active tickets are closed.
- Both staff and non-staff have access to a ticket panel listing tickets, their statuses, and their messages. Non-staff can only see their own tickets. This panel is optional and all features are available inline with chat.

I think that covers everything.

### CODE REVIEW

File: code/__defines/_macros.dm [hunk 1]
```diff
@@ -74,9 +76,12 @@
...
 #define send_rsc(target, rsc_content, rsc_name)             target << browse_rsc(rsc_content, rsc_name)
+#define send_output(target, msg, control)                   target << output(msg, control)
 
 #define CanInteract(user, state) (CanUseTopic(user, state) == STATUS_INTERACTIVE)
 
 #define isopenturf(target) istype(target, /turf/simulated/open)
 #define isweakref(target) istype(target, /datum/weakref)
 #define isdatum(target) istype(target, /datum)
+
+#define any2ref(x) "\ref[x]"
```
Comment (Lohikar) [inline code/__defines/_macros.dm]: We have this as `SOFTREF()` already.

File: code/datums/repositories/client.dm [hunk 2]
```diff
@@ -0,0 +1,61 @@
+var/const/NO_CLIENT_CKEY = "*no ckey*"
+
+var/global/repository/client/client_repository = new()
```
Comment (Lohikar) [inline code/datums/repositories/client.dm]: What is this for?

File: code/datums/repositories/mobs.dm [hunk 3]
```diff
@@ -0,0 +1,28 @@
+var/global/repository/mob/mob_repository = new()
```
Comment (Lohikar) [inline code/datums/repositories/mobs.dm]: What is this for?

**Comment by mustafakalash** [comment]
Lite clients and mobs allow references to clients without fear of them nulling on logout.

**Comment by skull132** [comment]
I'll toss this up to the Head Memes for review. But chances are this'll be denied, as all discussions we've had about this have ended with, "Why is this convoluted as fuck?"

File: code/datums/repositories/client.dm (hunk 2 above)
Comment (skull132) [reply code/datums/repositories/client.dm]: Psure it's a global list of lite clients or whatever.

File: code/datums/repositories/client.dm [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+	return pm_link ? "<a href='?priv_msg=[ref];ticket=\ref[ticket]'>[key]</a>/([name])[rank2text()]" : "[key]/([name])"
+
+/datum/client_lite/proc/rank2text()
+	var/client/C = client_by_ckey(ckey)
+	if(!C || (C && !C.holder))
+		return
+	return " \[[C.holder.rank]\]"
+
+/proc/client_by_ckey(ckey)
+	for(var/client/C)
```
Comment (skull132) [inline code/datums/repositories/client.dm]: No global loops please. They freeze the server for 2 seconds per loop, due to the size of our game world.

There's a clients global list somewhere, crawl that. Could it possibly have been "players"? Also, make this into a typeless loop.

**Comment by mustafakalash** [comment]
I've received more positive reviews than not from Bay staff about the implementation. Let me know if you're interested and I'll keep working on it then.

**Comment by skull132** [comment]
Bay staff operate _way_ differently from ours. Anyways, I'll have an answer within a day or so, so sounds good. DNM until then.

File: code/modules/admin/verbs/adminpm.dm [hunk 5]
```diff
@@ -46,6 +46,25 @@
...
 		return
 
+	if (src.handle_spam_prevention(msg,MUTE_ADMINHELP))
+		return
+
+	var/recieve_pm_type = "Player"
+	if(holder)
+		//mod PMs are maroon
+		//PMs sent from admins and mods display their rank
+		if(holder)
```
Comment (skull132) [inline code/modules/admin/verbs/adminpm.dm]:
```dm
if (holder)
	if (holder)
```

...?

File: code/modules/admin/verbs/adminpm.dm [hunk 6]
```diff
@@ -46,6 +46,25 @@
...
+			if(!C.holder && holder && holder.fakekey)
```
Comment (skull132) [inline code/modules/admin/verbs/adminpm.dm]: We've checked for `holder` twice already. This condenses to `if (!C.holder && holder.fakekey)`

File: code/modules/admin/verbs/adminpm.dm [hunk 7]
```diff
@@ -46,6 +46,25 @@
...
+				recieve_pm_type = "Admin"
+			else
+				recieve_pm_type = holder.rank
+
+	else if(C && !C.holder)
```
Comment (skull132) [inline code/modules/admin/verbs/adminpm.dm]: `C` is checked as `client` on line 44. This condenses to `else if (!C.holder)`

File: code/modules/admin/verbs/adminpm.dm [hunk 8]
```diff
@@ -46,6 +46,25 @@
...
+		src << "<span class='warning'>Error: Admin-PM: Non-admin to non-admin PM communication is forbidden.</span>"
+		return
+
+	msg = sanitize(msg)
```
Comment (skull132) [inline code/modules/admin/verbs/adminpm.dm]:
```dm
if (!check_rights(R_SERVER|R_DEBUG, 0))
	msg = sanitize(msg)
```

Having admins be able to HTML inject for top keks is a part of the Aurora experience, I'm afraid.

**Comment by skull132** [comment]
@mustafakalash Well, the madmins want to give this a try. So I'll lift the DNM.

**Comment by mustafakalash** [comment]
Okay, I'll get to work on it when I can. Some of these issues are just the result of going from repo to another and copying some of the bad code bay has for its pms. I'll rewrite without the repositories.

File: code/modules/admin/ticket.dm [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+				dat += "<p>No messages to display.</p></div>"
+	else
+		dat += "<p>No tickets to display.</p>"
+	dat += "</body></html>"
+
+	return jointext(dat, null)
+
+/datum/ticket_panel/Topic(href, href_list)
+	if(usr && usr != ticket_panel_window.user)
+		if(href_list["close"]) // catch the case where a user switches mobs, then closes the window that was linked to the old mob
```
[the remark this answers is not in the document]
Comment (mustafakalash) [reply code/modules/admin/ticket.dm]: Don't think I can as a limitation of byond.

File: code/modules/admin/ticket.dm [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+				dat += "<ul>[jointext(msg_dat, null)]</ul></div>"
+			else
+				dat += "<p>No messages to display.</p></div>"
+	else
+		dat += "<p>No tickets to display.</p>"
+	dat += "</body></html>"
+
+	return jointext(dat, null)
+
+/datum/ticket_panel/Topic(href, href_list)
```
[the remark this answers is not in the document]
Comment (mustafakalash) [reply code/modules/admin/ticket.dm]: Every one always has access to the panel. Non-staff only see their own tickets.

File: code/modules/admin/verbs/adminpm.dm (hunk 5 above)
Comment (mustafakalash) [reply code/modules/admin/verbs/adminpm.dm]: Lol this code has been on Bay for the longest time. I thought I had fixed it when I first wrote tickets.

File: code/modules/admin/verbs/adminpm.dm (hunk 8 above)
Comment (mustafakalash) [reply code/modules/admin/verbs/adminpm.dm]: It was the party poopers who removed it from Bay not long ago. 😞

**Comment by mustafakalash** [comment]
Testing looked good.

**Comment by skull132** [comment]
Will review tomorrow night or Tuesday.

**Comment by Arrow768** [comment]
Did some quick tests with Catnip and Juani.
Spotted a runtime and fixed it.
Other than that we couldn't find any issues with it.

**Pull request merged**